{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d937adc",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "pop_dir_tag = {\n",
    "    \"EstBB\":\"Est\",\"UKBB_AFR\":\"AFR\",\"UKBB_AMR\":\"AMR\",\"UKBB_CSA\":\"CSA\",\n",
    "    \"UKBB_EAS\":\"EAS\",\"UKBB_EUR\":\"EUR\",\"UKBB_MID\":\"MID\", \"meta_EUR\":\"meta_EUR\"\n",
    "}\n",
    "\n",
    "os.makedirs(\"annotated_coloc_results/FinnGen_lbf_coloc\", exist_ok=True)\n",
    "\n",
    "for key, value in pop_dir_tag.items():\n",
    "    metadata = pd.read_csv(f\"annotated_meta_summaries/{key}_summary_annotated.tsv\", sep=\"\\t\")\n",
    "\n",
    "    metadata = metadata[[\"signal\", \"metabolite\", \"maf\", \"log10p\", \"nearest_genes\", \"gene_symbols\"]].drop_duplicates()\n",
    "\n",
    "    metadata = metadata.rename(columns={\"signal\":\"signal2\", \"maf\": \"maf2\", \"log10p\": \"log10p2\"})\n",
    "\n",
    "    fg_metadata = pd.read_csv(f\"finngen_all_variants_maf_p.tsv\", sep=\"\\t\")\n",
    "    fg_metadata = fg_metadata.rename(columns={\"signal\":\"signal1\", \"maf\": \"maf1\"})\n",
    "    fg_metadata[\"log10p1\"] = -np.log10(fg_metadata[\"p\"])\n",
    "    fg_metadata = fg_metadata[[\"signal1\", \"log10p1\", \"maf1\", \"diagnosis\"]]\n",
    "\n",
    "    df = pd.read_csv(f\"meta_colocid/FinnGen_lbf_coloc/{key}_FinnGen_lbfs_results.tsv\", sep=\"\\t\")\n",
    "\n",
    "    df = df[df[\"PP.H4\"]>=0.8]\n",
    "\n",
    "    df = df.merge(metadata, on=(\"signal2\"), how=\"left\")\n",
    "    df = df.merge(fg_metadata, on=(\"signal1\"), how=\"left\")\n",
    "\n",
    "    # df[\"diagnosis\"] = df[\"signal1\"].str.replace(r\"_chr.*\", \"\", regex=True)\n",
    "\n",
    "    df.drop_duplicates(inplace=True)\n",
    "\n",
    "    df.to_csv(f\"annotated_coloc_results/FinnGen_lbf_coloc/{key}_FinnGen_lbf_coloc_results.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a43dd74",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/pandas/core/arraylike.py:399: RuntimeWarning: divide by zero encountered in log10\n",
      "  result = getattr(ufunc, method)(*inputs, **kwargs)\n",
      "/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/pandas/core/arraylike.py:399: RuntimeWarning: divide by zero encountered in log10\n",
      "  result = getattr(ufunc, method)(*inputs, **kwargs)\n",
      "/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/pandas/core/arraylike.py:399: RuntimeWarning: divide by zero encountered in log10\n",
      "  result = getattr(ufunc, method)(*inputs, **kwargs)\n",
      "/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/pandas/core/arraylike.py:399: RuntimeWarning: divide by zero encountered in log10\n",
      "  result = getattr(ufunc, method)(*inputs, **kwargs)\n",
      "/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/pandas/core/arraylike.py:399: RuntimeWarning: divide by zero encountered in log10\n",
      "  result = getattr(ufunc, method)(*inputs, **kwargs)\n",
      "/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/pandas/core/arraylike.py:399: RuntimeWarning: divide by zero encountered in log10\n",
      "  result = getattr(ufunc, method)(*inputs, **kwargs)\n",
      "/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/pandas/core/arraylike.py:399: RuntimeWarning: divide by zero encountered in log10\n",
      "  result = getattr(ufunc, method)(*inputs, **kwargs)\n",
      "/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/pandas/core/arraylike.py:399: RuntimeWarning: divide by zero encountered in log10\n",
      "  result = getattr(ufunc, method)(*inputs, **kwargs)\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "pop_dir_tag = {\n",
    "    \"EstBB\":\"Est\",\"UKBB_AFR\":\"AFR\",\"UKBB_AMR\":\"AMR\",\"UKBB_CSA\":\"CSA\",\n",
    "    \"UKBB_EAS\":\"EAS\",\"UKBB_EUR\":\"EUR\",\"UKBB_MID\":\"MID\", \"meta_EUR\":\"meta_EUR\"\n",
    "}\n",
    "\n",
    "os.makedirs(\"annotated_coloc_results/FinnGen_abf_coloc\", exist_ok=True)\n",
    "\n",
    "for key, value in pop_dir_tag.items():\n",
    "    metadata = pd.read_csv(f\"annotated_meta_summaries/{key}_summary_annotated.tsv\", sep=\"\\t\")\n",
    "\n",
    "    metadata = metadata[[\"signal\", \"metabolite\", \"maf\", \"log10p\", \"nearest_genes\", \"gene_symbols\"]].drop_duplicates()\n",
    "\n",
    "    metadata = metadata.rename(columns={\"signal\":\"signal2\", \"maf\": \"maf2\", \"log10p\": \"log10p2\"})\n",
    "\n",
    "    fg_metadata = pd.read_csv(f\"finngen_all_variants_maf_p.tsv\", sep=\"\\t\")\n",
    "    fg_metadata = fg_metadata.rename(columns={\"signal\":\"signal1\", \"maf\": \"maf1\"})\n",
    "    fg_metadata[\"log10p1\"] = -np.log10(fg_metadata[\"p\"])\n",
    "    fg_metadata = fg_metadata[[\"signal1\", \"log10p1\", \"maf1\", \"diagnosis\"]]\n",
    "\n",
    "    df = pd.read_csv(f\"meta_colocid/FinnGen_abf_coloc/{key}_FinnGen_abfs_results.tsv\", sep=\"\\t\")\n",
    "\n",
    "    df = df[df[\"PP.H4\"]>=0.8]\n",
    "\n",
    "    df = df.merge(metadata, on=(\"signal2\"), how=\"left\")\n",
    "    df = df.merge(fg_metadata, on=(\"signal1\"), how=\"left\")\n",
    "\n",
    "    # df[\"diagnosis\"] = df[\"signal1\"].str.replace(r\"_chr.*\", \"\", regex=True)\n",
    "\n",
    "    df.drop_duplicates(inplace=True)\n",
    "\n",
    "    df.to_csv(f\"annotated_coloc_results/FinnGen_abf_coloc/{key}_FinnGen_abf_coloc_results.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5673f958",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/g4/mx686c290ms3_t_t0prfdknc0000gn/T/ipykernel_81380/4095574367.py:21: DtypeWarning: Columns (7) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  meta_metadata = pd.read_csv(f\"FinnGen+MVP+UKBB_important_variants.tsv\", sep=\"\\t\")\n",
      "/var/folders/g4/mx686c290ms3_t_t0prfdknc0000gn/T/ipykernel_81380/4095574367.py:21: DtypeWarning: Columns (7) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  meta_metadata = pd.read_csv(f\"FinnGen+MVP+UKBB_important_variants.tsv\", sep=\"\\t\")\n",
      "/var/folders/g4/mx686c290ms3_t_t0prfdknc0000gn/T/ipykernel_81380/4095574367.py:21: DtypeWarning: Columns (7) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  meta_metadata = pd.read_csv(f\"FinnGen+MVP+UKBB_important_variants.tsv\", sep=\"\\t\")\n",
      "/var/folders/g4/mx686c290ms3_t_t0prfdknc0000gn/T/ipykernel_81380/4095574367.py:21: DtypeWarning: Columns (7) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  meta_metadata = pd.read_csv(f\"FinnGen+MVP+UKBB_important_variants.tsv\", sep=\"\\t\")\n",
      "/var/folders/g4/mx686c290ms3_t_t0prfdknc0000gn/T/ipykernel_81380/4095574367.py:21: DtypeWarning: Columns (7) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  meta_metadata = pd.read_csv(f\"FinnGen+MVP+UKBB_important_variants.tsv\", sep=\"\\t\")\n",
      "/var/folders/g4/mx686c290ms3_t_t0prfdknc0000gn/T/ipykernel_81380/4095574367.py:21: DtypeWarning: Columns (7) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  meta_metadata = pd.read_csv(f\"FinnGen+MVP+UKBB_important_variants.tsv\", sep=\"\\t\")\n",
      "/var/folders/g4/mx686c290ms3_t_t0prfdknc0000gn/T/ipykernel_81380/4095574367.py:21: DtypeWarning: Columns (7) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  meta_metadata = pd.read_csv(f\"FinnGen+MVP+UKBB_important_variants.tsv\", sep=\"\\t\")\n",
      "/var/folders/g4/mx686c290ms3_t_t0prfdknc0000gn/T/ipykernel_81380/4095574367.py:21: DtypeWarning: Columns (7) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  meta_metadata = pd.read_csv(f\"FinnGen+MVP+UKBB_important_variants.tsv\", sep=\"\\t\")\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import numpy as np\n",
    "\n",
    "pop_dir_tag = {\n",
    "    \"EstBB\":\"Est\",\"UKBB_AFR\":\"AFR\",\"UKBB_AMR\":\"AMR\",\"UKBB_CSA\":\"CSA\",\n",
    "    \"UKBB_EAS\":\"EAS\",\"UKBB_EUR\":\"EUR\",\"UKBB_MID\":\"MID\", \"meta_EUR\":\"meta_EUR\"\n",
    "}\n",
    "\n",
    "os.makedirs(\"annotated_gpu-coloc_results/FinnGen+MVP+UKBB_coloc\", exist_ok=True)\n",
    "\n",
    "for key, value in pop_dir_tag.items():\n",
    "    metadata = pd.read_csv(f\"annotated_meta_summaries/{key}_summary_annotated.tsv\", sep=\"\\t\")\n",
    "\n",
    "    metadata = metadata[[\"signal\", \"metabolite\", \"maf\", \"log10p\", \"nearest_genes\", 'gene_symbols']].drop_duplicates()\n",
    "\n",
    "    metadata = metadata.rename(columns={\"signal\":\"signal2\", \"maf\": \"maf2\", \"log10p\": \"log10p2\"})\n",
    "\n",
    "    df = pd.read_csv(f\"meta_colocid/FinnGen+MVP+UKBB_coloc/{key}_FinnGen+MVP+UKBB_META_results.tsv\", sep=\"\\t\")\n",
    "\n",
    "    meta_metadata = pd.read_csv(f\"FinnGen+MVP+UKBB_important_variants.tsv\", sep=\"\\t\")\n",
    "\n",
    "    meta_metadata[\"log10p1\"] = -np.log10(meta_metadata[\"all_inv_var_meta_p\"])\n",
    "\n",
    "    meta_metadata[\"maf1_FinnGen\"] = meta_metadata[\"fg_af_alt\"].apply(lambda x: min(x, 1-x) if pd.notna(x) else None)\n",
    "    meta_metadata[\"maf1_MVP_EUR\"] = meta_metadata[\"MVP_EUR_af_alt\"].apply(lambda x: min(x, 1-x) if pd.notna(x) else None)\n",
    "\n",
    "    meta_metadata = meta_metadata.rename(columns={\"variant\": \"lead1\"})\n",
    "    \n",
    "    meta_metadata = meta_metadata[[\"log10p1\", \"maf1_FinnGen\", \"maf1_MVP_EUR\", \"trait\", \"lead1\"]]\n",
    "    \n",
    "    df = df[df[\"PP.H4\"]>=0.8]\n",
    "\n",
    "    df = df.merge(metadata, on=(\"signal2\"), how=\"left\")\n",
    "\n",
    "    df[\"trait\"] = (\n",
    "        df[\"signal1\"]\n",
    "        .str.replace(r\"_chr.*\", \"\", regex=True)\n",
    "        .str.removeprefix(\"FinnGen+MVP+UKBB_\")\n",
    "    )\n",
    "\n",
    "    df = df.merge(meta_metadata, on=(\"trait\", \"lead1\"), how=\"left\")\n",
    "\n",
    "    df.drop_duplicates(inplace=True)\n",
    "\n",
    "    df.to_csv(f\"annotated_gpu-coloc_results/FinnGen+MVP+UKBB_coloc/{value}_FinnGen+MVP+UKBB_coloc_results.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c65b86fa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(0, 26)\n",
      "(0, 26)\n",
      "(0, 26)\n",
      "(0, 26)\n",
      "(0, 26)\n",
      "(0, 26)\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "pop_dir_tag = {\n",
    "    \"EstBB\":\"EUR\",\"UKBB_AFR\":\"AFR\",\"UKBB_CSA\":\"CSA\",\n",
    "    \"UKBB_EAS\":\"EAS\",\"UKBB_EUR\":\"EUR\",\"UKBB_MID\":\"MID\"#, \"meta_EUR\":\"meta_EUR\"\n",
    "}\n",
    "\n",
    "os.makedirs(\"annotated_coloc_results/PANUKBB_coloc\", exist_ok=True)\n",
    "\n",
    "for key, value in pop_dir_tag.items():\n",
    "    metadata = pd.read_csv(f\"annotated_meta_summaries/{key}_summary_annotated.tsv\", sep=\"\\t\")\n",
    "\n",
    "    metadata = metadata[[\"signal\", \"metabolite\", \"maf\", \"log10p\", \"nearest_genes\", \"gene_symbols\"]].drop_duplicates()\n",
    "\n",
    "    metadata = metadata.rename(columns={\"signal\":\"signal2\", \"maf\":\"maf2\", \"log10p\": \"log10p2\"})\n",
    "    \n",
    "    df = pd.read_csv(f\"meta_colocid/PANUKBB_coloc/PANUKBB_{value}_{key}_results.tsv\", sep=\"\\t\")\n",
    "\n",
    "    df = df.merge(metadata, on=(\"signal2\"), how=\"left\")\n",
    "\n",
    "    panukbb_metadata = pd.read_csv(f\"APS/PANUKBB_{value}_annotated_summary.tsv\", sep=\"\\t\")\n",
    "\n",
    "    panukbb_metadata = panukbb_metadata.rename(columns={\"signal\":\"signal1\", \"maf\":\"maf1\", \"-log10p\": \"log10p1\"})\n",
    "\n",
    "    panukbb_metadata.drop(columns=[\"chromosome\", \"lead_variant\", \"location_min\",  \"location_max\",  \"signal_strength\", \"lead_variant\"])\n",
    "\n",
    "    df = df.merge(panukbb_metadata, on=\"signal1\", how=\"left\")\n",
    "    print(df[df[\"trait_type\"].isna()].shape)\n",
    "    # df.dropna(subset=[\"trait_type\"], inplace=True)\n",
    "\n",
    "    df.drop_duplicates(inplace=True)\n",
    "\n",
    "    df.to_csv(f\"annotated_coloc_results/PANUKBB_coloc/PANUKBB_{value}_{key}_results.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0b0c7e7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/g4/mx686c290ms3_t_t0prfdknc0000gn/T/ipykernel_52759/3269977408.py:27: DtypeWarning: Columns (11) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  panukbb_metadata = pd.read_csv(f\"APS/PANUKBB_META_HQ_annotated_summary.tsv\", sep=\"\\t\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(0, 26)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/g4/mx686c290ms3_t_t0prfdknc0000gn/T/ipykernel_52759/3269977408.py:27: DtypeWarning: Columns (11) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  panukbb_metadata = pd.read_csv(f\"APS/PANUKBB_META_HQ_annotated_summary.tsv\", sep=\"\\t\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(0, 26)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/g4/mx686c290ms3_t_t0prfdknc0000gn/T/ipykernel_52759/3269977408.py:27: DtypeWarning: Columns (11) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  panukbb_metadata = pd.read_csv(f\"APS/PANUKBB_META_HQ_annotated_summary.tsv\", sep=\"\\t\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(0, 26)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/g4/mx686c290ms3_t_t0prfdknc0000gn/T/ipykernel_52759/3269977408.py:27: DtypeWarning: Columns (11) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  panukbb_metadata = pd.read_csv(f\"APS/PANUKBB_META_HQ_annotated_summary.tsv\", sep=\"\\t\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(0, 26)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/g4/mx686c290ms3_t_t0prfdknc0000gn/T/ipykernel_52759/3269977408.py:27: DtypeWarning: Columns (11) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  panukbb_metadata = pd.read_csv(f\"APS/PANUKBB_META_HQ_annotated_summary.tsv\", sep=\"\\t\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(0, 26)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/g4/mx686c290ms3_t_t0prfdknc0000gn/T/ipykernel_52759/3269977408.py:27: DtypeWarning: Columns (11) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  panukbb_metadata = pd.read_csv(f\"APS/PANUKBB_META_HQ_annotated_summary.tsv\", sep=\"\\t\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(0, 26)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/g4/mx686c290ms3_t_t0prfdknc0000gn/T/ipykernel_52759/3269977408.py:27: DtypeWarning: Columns (11) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  panukbb_metadata = pd.read_csv(f\"APS/PANUKBB_META_HQ_annotated_summary.tsv\", sep=\"\\t\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(0, 26)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/g4/mx686c290ms3_t_t0prfdknc0000gn/T/ipykernel_52759/3269977408.py:27: DtypeWarning: Columns (11) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  panukbb_metadata = pd.read_csv(f\"APS/PANUKBB_META_HQ_annotated_summary.tsv\", sep=\"\\t\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(0, 26)\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "pop_dir_tag = {\n",
    "    \"EstBB\":\"EstBB\",\"UKBB_AFR\":\"AFR\",\"UKBB_AMR\":\"AMR\",\"UKBB_CSA\":\"CSA\",\n",
    "    \"UKBB_EAS\":\"EAS\",\"UKBB_EUR\":\"EUR\",\"UKBB_MID\":\"MID\", \"meta_EUR\":\"meta_EUR\"\n",
    "}\n",
    "\n",
    "os.makedirs(\"annotated_coloc_results/PANUKBB_coloc\", exist_ok=True)\n",
    "\n",
    "for key, value in pop_dir_tag.items():\n",
    "    metadata = pd.read_csv(f\"annotated_meta_summaries/{key}_summary_annotated.tsv\", sep=\"\\t\")\n",
    "\n",
    "    metadata = metadata[[\"signal\", \"metabolite\", \"maf\", \"log10p\", \"nearest_genes\", \"gene_symbols\"]].drop_duplicates()\n",
    "\n",
    "    metadata = metadata.rename(columns={\"signal\":\"signal2\", \"maf\": \"maf2\", \"log10p\": \"log10p2\"})\n",
    "\n",
    "    #MID_continuous-102540-both_sexes_chr7:92732610-94732002\n",
    "    panukbb_metadata = pd.read_csv(f\"PAN-UK_BB.manifest.csv\")\n",
    "\n",
    "    df = pd.read_csv(f\"meta_colocid/PANUKBB_coloc/PANUKBB_META_HQ_{key}_results.tsv\", sep=\"\\t\")\n",
    "\n",
    "    df = df[df[\"PP.H4\"]>=0.8]\n",
    "\n",
    "    df = df.merge(metadata, on=(\"signal2\"), how=\"left\")\n",
    "\n",
    "    panukbb_metadata = pd.read_csv(f\"APS/PANUKBB_META_HQ_annotated_summary.tsv\", sep=\"\\t\")\n",
    "\n",
    "    panukbb_metadata = panukbb_metadata.rename(columns={\"signal\":\"signal1\", \"maf\":\"maf1\", \"-log10p\": \"log10p1\"})\n",
    "\n",
    "    panukbb_metadata.drop(columns=[\"chromosome\", \"lead_variant\", \"location_min\",  \"location_max\",  \"signal_strength\", \"lead_variant\"])\n",
    "\n",
    "    df = df.merge(panukbb_metadata, on=\"signal1\", how=\"left\")\n",
    "    print(df[df[\"trait_type\"].isna()].shape)\n",
    "    df.drop_duplicates(inplace=True)\n",
    "\n",
    "    df.to_csv(f\"annotated_coloc_results/PANUKBB_coloc/PANUKBB_META_HQ_{key}_results.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "626258d2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(0, 26)\n",
      "(0, 26)\n",
      "(0, 26)\n",
      "(0, 26)\n",
      "(0, 26)\n",
      "(0, 26)\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "pop_dir_tag = {\n",
    "   \"UKBB_AFR\":\"AFR\",\"UKBB_AMR\":\"AMR\",\"UKBB_CSA\":\"CSA\",\n",
    "    \"UKBB_EAS\":\"EAS\",\"UKBB_EUR\":\"EUR\",\"UKBB_MID\":\"MID\"#, \"meta_EUR\":\"meta_EUR\"\n",
    "}\n",
    "\n",
    "os.makedirs(\"annotated_coloc_results/PANUKBB_coloc\", exist_ok=True)\n",
    "\n",
    "for key, value in pop_dir_tag.items():\n",
    "    metadata = pd.read_csv(f\"annotated_meta_summaries/meta_EUR_summary_annotated.tsv\", sep=\"\\t\")\n",
    "\n",
    "    metadata = metadata[[\"signal\", \"metabolite\", \"maf\", \"log10p\", \"nearest_genes\", \"gene_symbols\"]].drop_duplicates()\n",
    "\n",
    "    metadata = metadata.rename(columns={\"signal\":\"signal2\", \"maf\": \"maf2\", \"log10p\": \"log10p2\"})\n",
    "\n",
    "    #MID_continuous-102540-both_sexes_chr7:92732610-94732002\n",
    "    panukbb_metadata = pd.read_csv(f\"PAN-UK_BB.manifest.csv\")\n",
    "\n",
    "    df = pd.read_csv(f\"meta_colocid/PANUKBB_coloc/PANUKBB_{value}_meta_EUR_results.tsv\", sep=\"\\t\")\n",
    "\n",
    "    df = df[df[\"PP.H4\"]>=0.8]\n",
    "\n",
    "    df = df.merge(metadata, on=(\"signal2\"), how=\"left\")\n",
    "\n",
    "    panukbb_metadata = pd.read_csv(f\"APS/PANUKBB_{value}_annotated_summary.tsv\", sep=\"\\t\")\n",
    "\n",
    "    panukbb_metadata = panukbb_metadata.rename(columns={\"signal\":\"signal1\", \"maf\":\"maf1\", \"-log10p\": \"log10p1\"})\n",
    "\n",
    "    panukbb_metadata.drop(columns=[\"chromosome\", \"lead_variant\", \"location_min\",  \"location_max\",  \"signal_strength\", \"lead_variant\"])\n",
    "\n",
    "    df = df.merge(panukbb_metadata, on=\"signal1\", how=\"left\")\n",
    "    print(df[df[\"trait_type\"].isna()].shape)\n",
    "\n",
    "    df.drop_duplicates(inplace=True)\n",
    "\n",
    "    df.to_csv(f\"annotated_coloc_results/PANUKBB_coloc/PANUKBB_{value}_meta_EUR_results.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c69c2386",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['signal', 'chromosome', 'location_min', 'location_max',\n",
      "       'signal_strength', 'lead_variant', 'chr', 'pos', 'trait', 'metabolite',\n",
      "       'maf', 'log10p', 'nearest_genes', 'gene_symbols'],\n",
      "      dtype='object')\n",
      "549\n",
      "549\n",
      "Index(['signal', 'chromosome', 'location_min', 'location_max',\n",
      "       'signal_strength', 'lead_variant', 'chr', 'pos', 'trait', 'metabolite',\n",
      "       'maf', 'log10p', 'nearest_genes', 'gene_symbols'],\n",
      "      dtype='object')\n",
      "105\n",
      "105\n",
      "Index(['signal', 'chromosome', 'location_min', 'location_max',\n",
      "       'signal_strength', 'lead_variant', 'chr', 'pos', 'trait', 'metabolite',\n",
      "       'maf', 'log10p', 'nearest_genes', 'gene_symbols'],\n",
      "      dtype='object')\n",
      "2496\n",
      "2496\n",
      "Index(['signal', 'chromosome', 'location_min', 'location_max',\n",
      "       'signal_strength', 'lead_variant', 'chr', 'pos', 'trait', 'metabolite',\n",
      "       'maf', 'log10p', 'nearest_genes', 'gene_symbols'],\n",
      "      dtype='object')\n",
      "1475\n",
      "1475\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "pop_dir_tag = {\n",
    "   \"UKBB_AFR\":\"AFR\",\"UKBB_CSA\":\"CSA\",\n",
    "    \"UKBB_EUR\":\"EUR\", \"EstBB\": \"EUR\" #\"meta_EUR\":\"meta_EUR\"\n",
    "}\n",
    "\n",
    "os.makedirs(\"annotated_coloc_results/UKBPPP_coloc\", exist_ok=True)\n",
    "\n",
    "for key, value in pop_dir_tag.items():\n",
    "    metadata = pd.read_csv(f\"annotated_meta_summaries/{key}_summary_annotated.tsv\", sep=\"\\t\")\n",
    "\n",
    "    metadata = metadata[[\"signal\", \"metabolite\", \"maf\", \"log10p\", \"nearest_genes\", \"gene_symbols\"]].drop_duplicates()\n",
    "\n",
    "    metadata = metadata.rename(columns={\"signal\":\"signal2\", \"maf\": \"maf2\", \"log10p\": \"log10p2\"})\n",
    "\n",
    "    df = pd.read_csv(f\"meta_colocid/UKBPPP_coloc/UKBPPP_{value}_{key}_results.tsv\", sep=\"\\t\")\n",
    "\n",
    "    df = df[df[\"PP.H4\"]>=0.8]\n",
    "\n",
    "    df = df.merge(metadata, on=(\"signal2\"), how=\"left\")\n",
    "\n",
    "    df[\"UKBPPP_ProteinID\"] = df[\"signal1\"].str.split(\"_chr\").str[1].str.split(\"_\").str[1].str.split(\":\").str[0:-1].str.join(\":\")\n",
    "\n",
    "    prot_met = pd.read_csv(\"Protein annotation/olink_protein_map_3k_v1.tsv\", sep=\"\\t\", usecols=[\"UKBPPP_ProteinID\", \"olink_target_fullname\", \"OlinkID\", \"UniProt\", \"Assay\", \"UniProt2\", \"HGNC.symbol\", \"Panel\"])\n",
    "\n",
    "    print(df.shape[0])\n",
    "\n",
    "    df = df.merge(prot_met, on=\"UKBPPP_ProteinID\", how=\"left\")\n",
    "\n",
    "    print(df.shape[0])\n",
    "\n",
    "    df.drop_duplicates(inplace=True)\n",
    "\n",
    "    df.to_csv(f\"annotated_coloc_results/UKBPPP_coloc/UKBPPP_{value}_{key}_results.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "782cff8e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "36\n",
      "36\n",
      "115\n",
      "115\n",
      "16\n",
      "16\n",
      "2904\n",
      "2904\n",
      "3567\n",
      "3567\n",
      "1523\n",
      "1523\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "pop_dir_tag = {\n",
    "   \"UKBB_AFR\":\"AFR\",\"UKBB_CSA\":\"CSA\", \"UKBB_EAS\": \"EAS\",\n",
    "    \"UKBB_EUR\":\"EUR\", \"meta_EUR\":\"meta_EUR\", \"EstBB\": \"EUR\"\n",
    "}\n",
    "\n",
    "os.makedirs(\"annotated_coloc_results/UKBPPP_coloc\", exist_ok=True)\n",
    "\n",
    "value = \"Combined\"\n",
    "\n",
    "for key, _ in pop_dir_tag.items():\n",
    "    metadata = pd.read_csv(f\"annotated_meta_summaries/{key}_summary_annotated.tsv\", sep=\"\\t\")\n",
    "\n",
    "    metadata = metadata[[\"signal\", \"metabolite\", \"maf\", \"log10p\", \"nearest_genes\", \"gene_symbols\"]].drop_duplicates()\n",
    "\n",
    "    metadata = metadata.rename(columns={\"signal\":\"signal2\", \"maf\": \"maf2\", \"log10p\": \"log10p2\"})\n",
    "\n",
    "    df = pd.read_csv(f\"meta_colocid/UKBPPP_coloc/UKBPPP_{value}_{key}_results.tsv\", sep=\"\\t\")\n",
    "\n",
    "    df = df[df[\"PP.H4\"]>=0.8]\n",
    "\n",
    "    df = df.merge(metadata, on=(\"signal2\"), how=\"left\")\n",
    "\n",
    "    df[\"UKBPPP_ProteinID\"] = df[\"signal1\"].str.split(\"_chr\").str[1].str.split(\"_\").str[1].str.split(\":\").str[0:-1].str.join(\":\")\n",
    "\n",
    "    prot_met = pd.read_csv(\"Protein annotation/olink_protein_map_3k_v1.tsv\", sep=\"\\t\", usecols=[\"UKBPPP_ProteinID\", \"olink_target_fullname\", \"OlinkID\", \"UniProt\", \"Assay\", \"UniProt2\", \"HGNC.symbol\", \"Panel\"])\n",
    "\n",
    "    print(df.shape[0])\n",
    "\n",
    "    df = df.merge(prot_met, on=\"UKBPPP_ProteinID\", how=\"left\")\n",
    "\n",
    "    print(df.shape[0])\n",
    "\n",
    "    df.drop_duplicates(inplace=True)\n",
    "\n",
    "    df.to_csv(f\"annotated_coloc_results/UKBPPP_coloc/UKBPPP_{value}_{key}_results.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7daf8afc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "965\n",
      "965\n",
      "1443\n",
      "1443\n",
      "3167\n",
      "3167\n",
      "624\n",
      "624\n",
      "3567\n",
      "3567\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "pop_dir_tag = {\n",
    "   \"UKBB_AFR\":\"AFR\",\"UKBB_CSA\":\"CSA\", \n",
    "    \"UKBB_EUR\":\"EUR\", \"UKBB_MID\": \"MID\", \"meta_EUR\": \"Combined\"\n",
    "}\n",
    "\n",
    "os.makedirs(\"annotated_coloc_results/UKBPPP_coloc\", exist_ok=True)\n",
    "\n",
    "key = \"meta_EUR\"\n",
    "\n",
    "for _, value in pop_dir_tag.items():\n",
    "    metadata = pd.read_csv(f\"annotated_meta_summaries/{key}_summary_annotated.tsv\", sep=\"\\t\")\n",
    "\n",
    "    metadata = metadata[[\"signal\", \"metabolite\", \"maf\", \"log10p\", \"nearest_genes\", \"gene_symbols\"]].drop_duplicates()\n",
    "\n",
    "    metadata = metadata.rename(columns={\"signal\":\"signal2\", \"maf\": \"maf2\", \"log10p\": \"log10p2\"})\n",
    "\n",
    "    df = pd.read_csv(f\"meta_colocid/UKBPPP_coloc/UKBPPP_{value}_{key}_results.tsv\", sep=\"\\t\")\n",
    "\n",
    "    df = df[df[\"PP.H4\"]>=0.8]\n",
    "\n",
    "    df = df.merge(metadata, on=(\"signal2\"), how=\"left\")\n",
    "\n",
    "    df[\"UKBPPP_ProteinID\"] = df[\"signal1\"].str.split(\"_chr\").str[1].str.split(\"_\").str[1].str.split(\":\").str[0:-1].str.join(\":\")\n",
    "\n",
    "    prot_met = pd.read_csv(\"Protein annotation/olink_protein_map_3k_v1.tsv\", sep=\"\\t\", usecols=[\"UKBPPP_ProteinID\", \"olink_target_fullname\", \"OlinkID\", \"UniProt\", \"Assay\", \"UniProt2\", \"HGNC.symbol\", \"Panel\"])\n",
    "\n",
    "    print(df.shape[0])\n",
    "\n",
    "    df = df.merge(prot_met, on=\"UKBPPP_ProteinID\", how=\"left\")\n",
    "\n",
    "    print(df.shape[0])\n",
    "\n",
    "    df.drop_duplicates(inplace=True)\n",
    "\n",
    "    df.to_csv(f\"annotated_coloc_results/UKBPPP_coloc/UKBPPP_{value}_{key}_results.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8fca185d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "109\n",
      "109\n",
      "0\n",
      "0\n",
      "55\n",
      "55\n",
      "18\n",
      "18\n",
      "15564\n",
      "15564\n",
      "1\n",
      "1\n",
      "20429\n",
      "20429\n",
      "5044\n",
      "5044\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "pop_dir_tag = {\n",
    "   \"UKBB_AFR\":\"AFR\",\"UKBB_AMR\":\"AMR\",\"UKBB_CSA\":\"CSA\",\n",
    "    \"UKBB_EAS\":\"EAS\",\"UKBB_EUR\":\"EUR\",\"UKBB_MID\":\"MID\", \"meta_EUR\":\"meta_EUR\",\n",
    "    \"EstBB\": \"EstBB\"\n",
    "}\n",
    "\n",
    "os.makedirs(\"annotated_coloc_results/INTERVAL_coloc\", exist_ok=True)\n",
    "\n",
    "ensemble = pd.read_csv(\"ensemble.tsv\", sep=\"\\t\")\n",
    "\n",
    "for key, value in pop_dir_tag.items():\n",
    "    metadata = pd.read_csv(f\"annotated_meta_summaries/{key}_summary_annotated.tsv\", sep=\"\\t\")\n",
    "\n",
    "    metadata = metadata[[\"signal\", \"metabolite\", \"maf\", \"log10p\", \"nearest_genes\", \"gene_symbols\"]].drop_duplicates()\n",
    "\n",
    "    metadata = metadata.rename(columns={\"signal\":\"signal2\", \"maf\": \"maf2\", \"log10p\": \"log10p2\"})\n",
    "\n",
    "    df = pd.read_csv(f\"meta_colocid/INTERVAL_coloc/INTERVAL_{key}_results.tsv\", sep=\"\\t\")\n",
    "\n",
    "    df = df[df[\"PP.H4\"]>=0.8]\n",
    "\n",
    "    df[\"molecular_trait_id\"] = df[\"signal1\"].str.extract(r\"(ENSG\\d{11})\")\n",
    "\n",
    "    df = df.merge(metadata, on=(\"signal2\"), how=\"left\")\n",
    "\n",
    "    print(df.shape[0])\n",
    "\n",
    "    df = df.merge(ensemble, on=\"molecular_trait_id\", how=\"left\")\n",
    "\n",
    "    print(df.shape[0])\n",
    "\n",
    "    df.drop_duplicates(inplace=True)\n",
    "\n",
    "    df.to_csv(f\"annotated_coloc_results/INTERVAL_coloc/INTERVAL_{key}_results.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb717dbc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "840\n",
      "840\n",
      "991\n",
      "991\n",
      "211\n",
      "538\n",
      "307\n",
      "307\n",
      "677\n",
      "677\n",
      "3\n",
      "3\n",
      "10\n",
      "10\n",
      "7\n",
      "7\n",
      "8\n",
      "8\n",
      "851\n",
      "851\n",
      "1567\n",
      "1567\n",
      "200\n",
      "550\n",
      "209\n",
      "209\n",
      "359\n",
      "359\n",
      "209\n",
      "209\n",
      "382\n",
      "382\n",
      "21\n",
      "51\n",
      "55\n",
      "55\n",
      "148\n",
      "148\n",
      "150842\n",
      "150842\n",
      "170832\n",
      "170832\n",
      "52055\n",
      "154296\n",
      "78954\n",
      "78954\n",
      "89526\n",
      "89526\n",
      "63\n",
      "63\n",
      "81\n",
      "81\n",
      "10\n",
      "20\n",
      "31\n",
      "31\n",
      "38\n",
      "38\n",
      "199084\n",
      "199084\n",
      "221732\n",
      "221732\n",
      "76534\n",
      "235935\n",
      "111673\n",
      "111673\n",
      "121723\n",
      "121723\n",
      "54711\n",
      "54711\n",
      "53389\n",
      "53389\n",
      "24132\n",
      "84665\n",
      "36027\n",
      "36027\n",
      "31617\n",
      "31617\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "pop_dir_tag = {\n",
    "   \"UKBB_AFR\":\"AFR\",\"UKBB_AMR\":\"AMR\",\"UKBB_CSA\":\"CSA\",\n",
    "    \"UKBB_EAS\":\"EAS\",\"UKBB_EUR\":\"EUR\",\"UKBB_MID\":\"MID\", \"meta_EUR\":\"meta_EUR\",\n",
    "    \"EstBB\": \"EstBB\"\n",
    "}\n",
    "\n",
    "os.makedirs(\"annotated_coloc_results/eQTL_Catalogue_coloc\", exist_ok=True)\n",
    "\n",
    "ensemble = pd.read_csv(\"ensemble.tsv\", sep=\"\\t\")\n",
    "\n",
    "eQTLC_metadata = pd.read_csv(\"tabix_ftp_paths.tsv\", sep=\"\\t\")\n",
    "\n",
    "eQTLC_metadata = eQTLC_metadata.rename(columns={\n",
    "    'dataset_id': 'eqtl_catalogue_dataset_id',\n",
    "    'quant_method': 'eqtl_catalogue_quant_method'\n",
    "})\n",
    "\n",
    "dataset_to_quant = dict(zip(eQTLC_metadata[\"eqtl_catalogue_dataset_id\"], eQTLC_metadata[\"eqtl_catalogue_quant_method\"]))\n",
    "\n",
    "ensemble = ensemble.rename(columns={\n",
    "    'molecular_trait_id': \"eqtl_catalogue_molecular_trait_id\",\n",
    "    'gene_id': 'eqtl_catalogue_gene_id',\n",
    "    'gene_name': 'eqtl_catalogue_gene_name'\n",
    "})\n",
    "\n",
    "for key, value in pop_dir_tag.items():\n",
    "    for method in [\"exon\", \"ge_microarray_aptamer\", \"leafcutter\", \"tx\", \"txrev\"]:\n",
    "        try:\n",
    "\n",
    "            metadata = pd.read_csv(f\"annotated_meta_summaries/{key}_summary_annotated.tsv\", sep=\"\\t\")\n",
    "\n",
    "            metadata = metadata[[\"signal\", \"metabolite\", \"maf\", \"log10p\", \"nearest_genes\", \"gene_symbols\"]].drop_duplicates()\n",
    "\n",
    "            metadata = metadata.rename(columns={\"signal\":\"signal2\", \"maf\": \"maf2\", \"log10p\": \"log10p2\"})\n",
    "\n",
    "            df = pd.read_csv(f\"meta_colocid/eQTL_catalogue_results/{method}_{key}_results.tsv\", sep=\"\\t\")\n",
    "\n",
    "            df = df[df[\"PP.H4\"]>=0.8]\n",
    "\n",
    "            df = df.merge(metadata, on=(\"signal2\"), how=\"left\")\n",
    "\n",
    "            print(df.shape[0])\n",
    "\n",
    "            df[\"eqtl_catalogue_quant_method\"] = df[\"signal1\"].str.split(\"_\").str[0].map(dataset_to_quant)\n",
    "            df[\"eqtl_catalogue_molecular_trait_id\"] = df[\"signal1\"].str.extract(r\"^QTD[^_]*_(.*?)_L\")\n",
    "\n",
    "            df = df.merge(ensemble, how=\"left\", on=\"eqtl_catalogue_molecular_trait_id\")\n",
    "\n",
    "            print(df.shape[0])\n",
    "\n",
    "            df.drop_duplicates(inplace=True)\n",
    "\n",
    "            df.to_csv(f\"annotated_coloc_results/eQTL_Catalogue_coloc/{method}_{key}_results.tsv\", sep=\"\\t\", index=False)    \n",
    "        except:\n",
    "            continue\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17e2eb68",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 5892/5892 [00:02<00:00, 2474.71it/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MVP_trait_id</th>\n",
       "      <th>MVP_trait</th>\n",
       "      <th>MVP_trait_description</th>\n",
       "      <th>MVP_pop_trait_sample_size</th>\n",
       "      <th>MVP_pop</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Phe_368</td>\n",
       "      <td>Visual disturbances</td>\n",
       "      <td>GWAS of Phe_368 (Visual disturbances) in Europ...</td>\n",
       "      <td>Total Sample Size=395764; Cases=50666; Control...</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Phe_187_8</td>\n",
       "      <td>Neoplasm of uncertain behavior of male genital...</td>\n",
       "      <td>GWAS of Phe_187_8 (Neoplasm of uncertain behav...</td>\n",
       "      <td>Total Sample Size=416269; Cases=594; Controls=...</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Phe_1005</td>\n",
       "      <td>Other symptoms</td>\n",
       "      <td>GWAS of Phe_1005 (Other symptoms) in Hispanic/...</td>\n",
       "      <td>Total Sample Size=50215; Cases=10590; Controls...</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Glucose_Mean_INT</td>\n",
       "      <td>glucose (fasting status unknown, mean, inv-nor...</td>\n",
       "      <td>Meta-analysis of Glucose_Mean_INT (glucose (fa...</td>\n",
       "      <td>Total Sample Size=601780</td>\n",
       "      <td>META</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PT_Min_INT</td>\n",
       "      <td>prothrombin time (PT, minimum, inv-norm transf...</td>\n",
       "      <td>Meta-analysis of PT_Min_INT (prothrombin time ...</td>\n",
       "      <td>Total Sample Size=120287</td>\n",
       "      <td>META</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5887</th>\n",
       "      <td>MomDoDM</td>\n",
       "      <td>Biological Mother: Diabetes / \"Sugar\"</td>\n",
       "      <td>GWAS of MomDoDM (Biological Mother: Diabetes /...</td>\n",
       "      <td>Total Sample Size=315668; Cases=46816; Control...</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5888</th>\n",
       "      <td>Glucose_FS_Min_INT</td>\n",
       "      <td>glucose (finger stick, minimum, inv-norm trans...</td>\n",
       "      <td>GWAS of Glucose_FS_Min_INT (glucose (finger st...</td>\n",
       "      <td>Total Sample Size=25853</td>\n",
       "      <td>AFR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5889</th>\n",
       "      <td>Phe_781</td>\n",
       "      <td>Symptoms involving nervous and musculoskeletal...</td>\n",
       "      <td>GWAS of Phe_781 (Symptoms involving nervous an...</td>\n",
       "      <td>Total Sample Size=117454; Cases=1661; Controls...</td>\n",
       "      <td>AFR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5890</th>\n",
       "      <td>HCT_Mean_INT</td>\n",
       "      <td>hematocrit (mean, inv-norm transformed)</td>\n",
       "      <td>Meta-analysis of HCT_Mean_INT (hematocrit (mea...</td>\n",
       "      <td>Total Sample Size=584645</td>\n",
       "      <td>META</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5891</th>\n",
       "      <td>Lymph_Fra_Max_INT</td>\n",
       "      <td>lymphocyte (fraction, maximum, inv-norm transf...</td>\n",
       "      <td>GWAS of Lymph_Fra_Max_INT (lymphocyte (fractio...</td>\n",
       "      <td>Total Sample Size=290958</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5892 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            MVP_trait_id                                          MVP_trait  \\\n",
       "0                Phe_368                                Visual disturbances   \n",
       "1              Phe_187_8  Neoplasm of uncertain behavior of male genital...   \n",
       "2               Phe_1005                                     Other symptoms   \n",
       "3       Glucose_Mean_INT  glucose (fasting status unknown, mean, inv-nor...   \n",
       "4             PT_Min_INT  prothrombin time (PT, minimum, inv-norm transf...   \n",
       "...                  ...                                                ...   \n",
       "5887             MomDoDM              Biological Mother: Diabetes / \"Sugar\"   \n",
       "5888  Glucose_FS_Min_INT  glucose (finger stick, minimum, inv-norm trans...   \n",
       "5889             Phe_781  Symptoms involving nervous and musculoskeletal...   \n",
       "5890        HCT_Mean_INT            hematocrit (mean, inv-norm transformed)   \n",
       "5891   Lymph_Fra_Max_INT  lymphocyte (fraction, maximum, inv-norm transf...   \n",
       "\n",
       "                                  MVP_trait_description  \\\n",
       "0     GWAS of Phe_368 (Visual disturbances) in Europ...   \n",
       "1     GWAS of Phe_187_8 (Neoplasm of uncertain behav...   \n",
       "2     GWAS of Phe_1005 (Other symptoms) in Hispanic/...   \n",
       "3     Meta-analysis of Glucose_Mean_INT (glucose (fa...   \n",
       "4     Meta-analysis of PT_Min_INT (prothrombin time ...   \n",
       "...                                                 ...   \n",
       "5887  GWAS of MomDoDM (Biological Mother: Diabetes /...   \n",
       "5888  GWAS of Glucose_FS_Min_INT (glucose (finger st...   \n",
       "5889  GWAS of Phe_781 (Symptoms involving nervous an...   \n",
       "5890  Meta-analysis of HCT_Mean_INT (hematocrit (mea...   \n",
       "5891  GWAS of Lymph_Fra_Max_INT (lymphocyte (fractio...   \n",
       "\n",
       "                              MVP_pop_trait_sample_size MVP_pop  \n",
       "0     Total Sample Size=395764; Cases=50666; Control...     EUR  \n",
       "1     Total Sample Size=416269; Cases=594; Controls=...     EUR  \n",
       "2     Total Sample Size=50215; Cases=10590; Controls...     AMR  \n",
       "3                              Total Sample Size=601780    META  \n",
       "4                              Total Sample Size=120287    META  \n",
       "...                                                 ...     ...  \n",
       "5887  Total Sample Size=315668; Cases=46816; Control...     EUR  \n",
       "5888                            Total Sample Size=25853     AFR  \n",
       "5889  Total Sample Size=117454; Cases=1661; Controls...     AFR  \n",
       "5890                           Total Sample Size=584645    META  \n",
       "5891                           Total Sample Size=290958     EUR  \n",
       "\n",
       "[5892 rows x 5 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from tqdm import tqdm\n",
    "\n",
    "MVP_traits = []\n",
    "\n",
    "for file in tqdm(os.listdir(\"MVP_metadata\")):\n",
    "    md = pd.read_csv(f\"MVP_metadata/{file}\", sep=\"\\t\")\n",
    "\n",
    "    wanted = [\"Title of analysis\",\"Analyzed variable\", \"Analysis description\", \"Sample size\"]\n",
    "\n",
    "    result = md.set_index(\"Info\").loc[wanted, \"Description\"].to_dict()\n",
    "    \n",
    "    result[\"MVP_trait_id\"] = result.pop(\"Title of analysis\").split(\".\")[0]\n",
    "    result[\"MVP_trait\"] = result.pop(\"Analyzed variable\")\n",
    "    result[\"MVP_trait_description\"] = result.pop(\"Analysis description\")\n",
    "    result[\"MVP_pop_trait_sample_size\"] = result.pop(\"Sample size\")\n",
    "    result[\"MVP_pop\"] = file.split(\".\")[3]\n",
    "\n",
    "    MVP_traits.append(result)\n",
    "\n",
    "MVP_traits_df = pd.DataFrame(MVP_traits)\n",
    "MVP_traits_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fe2d1af2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/14 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "43749\n",
      "43749\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  7%|▋         | 1/14 [00:03<00:41,  3.20s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "455171\n",
      "454741\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 14%|█▍        | 2/14 [00:24<02:44, 13.74s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "54947\n",
      "54947\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 21%|██▏       | 3/14 [00:26<01:31,  8.36s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "306562\n",
      "305674\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 29%|██▊       | 4/14 [00:41<01:52, 11.22s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "399453\n",
      "399059\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 36%|███▌      | 5/14 [00:56<01:52, 12.52s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "219681\n",
      "219483\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 50%|█████     | 7/14 [01:03<00:49,  7.05s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7320\n",
      "7320\n",
      "22\n",
      "22\n",
      "266399\n",
      "265635\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 64%|██████▍   | 9/14 [01:14<00:31,  6.31s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "156457\n",
      "155904\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 79%|███████▊  | 11/14 [01:19<00:13,  4.46s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8991\n",
      "7363\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 86%|████████▌ | 12/14 [01:19<00:06,  3.27s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "52\n",
      "52\n",
      "17009\n",
      "16894\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 14/14 [01:20<00:00,  5.74s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3089\n",
      "3089\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "from tqdm import tqdm\n",
    "\n",
    "pop_pairs = [\n",
    "            [\"meta_EUR\", \"AFR\"],[\"meta_EUR\", \"EUR\"],[\"meta_EUR\", \"AMR\"],[\"meta_EUR\", \"META\"],\n",
    "            [\"UKBB_EUR\", \"EUR\"],[\"EstBB\", \"EUR\"],[\"UKBB_AFR\", \"AFR\"],[\"UKBB_AMR\", \"AMR\"],\n",
    "            [\"UKBB_EUR\", \"META\"],[\"EstBB\", \"META\"],[\"UKBB_AFR\", \"META\"],[\"UKBB_AMR\", \"META\"],[\"UKBB_CSA\", \"META\"],[\"UKBB_EAS\", \"META\"],\n",
    "            ]\n",
    "\n",
    "os.makedirs(\"annotated_gpu-coloc_results/MVP_coloc\", exist_ok=True)\n",
    "\n",
    "for eur, mvp in tqdm(pop_pairs):\n",
    "    metadata = pd.read_csv(f\"annotated_meta_summaries/{eur}_summary_annotated.tsv\", sep=\"\\t\")\n",
    "\n",
    "    metadata = metadata[[\"signal\", \"metabolite\", \"maf\", \"log10p\", \"nearest_genes\", \"gene_symbols\"]].drop_duplicates()\n",
    "\n",
    "    metadata = metadata.rename(columns={\"signal\":\"signal2\", \"maf\": \"maf2\", \"log10p\": \"log10p2\"})\n",
    "\n",
    "    df = pd.read_csv(f\"MVP_coloc/MVP_{mvp}_{eur}_results.tsv\", sep=\"\\t\")\n",
    "\n",
    "    df = df[df[\"PP.H4\"]>=0.8]\n",
    "\n",
    "    df = df.merge(metadata, on=(\"signal2\"), how=\"left\")\n",
    "\n",
    "    df[\"MVP_trait_id\"] = df[\"signal1\"].str.split(\".\").str[2].str.split(\"_chr\").str[0]\n",
    "    df[\"MVP_pop\"] = df[\"signal1\"].str.split(\"_\").str[0]\n",
    "\n",
    "    MVP_traits_df = pd.read_csv(f\"annotated_MVP_summaries/{mvp}_summary.tsv\", sep=\"\\t\")\n",
    "\n",
    "    print(df.shape[0])\n",
    "\n",
    "    MVP_traits_df = MVP_traits_df.rename(columns={\"lead_variant1\": \"lead1\"})\n",
    "\n",
    "    df = df.merge(MVP_traits_df, on=(\"signal1\",\"MVP_trait_id\", \"MVP_pop\", \"lead1\"))\n",
    "\n",
    "    print(df.shape[0])\n",
    "\n",
    "    df.drop_duplicates(inplace=True)\n",
    "\n",
    "    df.to_csv(f\"annotated_gpu-coloc_results/MVP_coloc/MVP_{mvp}_{eur}_results.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "acceffe8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>lead1</th>\n",
       "      <th>study</th>\n",
       "      <th>log10p1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr9_22098620_A_G</td>\n",
       "      <td>GCST90132314_buildGRCh37</td>\n",
       "      <td>277.301899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr6_160589086_A_G</td>\n",
       "      <td>GCST90132314_buildGRCh37</td>\n",
       "      <td>185.661544</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr6_12903725_A_G</td>\n",
       "      <td>GCST90132314_buildGRCh37</td>\n",
       "      <td>113.395774</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1_109274968_G_T</td>\n",
       "      <td>GCST90132314_buildGRCh37</td>\n",
       "      <td>56.867740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr19_44908822_C_T</td>\n",
       "      <td>GCST90132314_buildGRCh37</td>\n",
       "      <td>53.983384</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>779</th>\n",
       "      <td>chr16_79536603_T_C</td>\n",
       "      <td>All_Metal_LDSC-CORR_Neff.v2</td>\n",
       "      <td>7.348916</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>780</th>\n",
       "      <td>chr2_141789968_C_T</td>\n",
       "      <td>All_Metal_LDSC-CORR_Neff.v2</td>\n",
       "      <td>7.380489</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>781</th>\n",
       "      <td>chr21_37150966_C_A</td>\n",
       "      <td>All_Metal_LDSC-CORR_Neff.v2</td>\n",
       "      <td>7.389021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>782</th>\n",
       "      <td>chr1_155348517_G_C</td>\n",
       "      <td>All_Metal_LDSC-CORR_Neff.v2</td>\n",
       "      <td>7.331335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>783</th>\n",
       "      <td>chr13_77413466_A_C</td>\n",
       "      <td>All_Metal_LDSC-CORR_Neff.v2</td>\n",
       "      <td>7.375512</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>977 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  lead1                        study     log10p1\n",
       "0     chr9_22098620_A_G     GCST90132314_buildGRCh37  277.301899\n",
       "1    chr6_160589086_A_G     GCST90132314_buildGRCh37  185.661544\n",
       "2     chr6_12903725_A_G     GCST90132314_buildGRCh37  113.395774\n",
       "3    chr1_109274968_G_T     GCST90132314_buildGRCh37   56.867740\n",
       "4    chr19_44908822_C_T     GCST90132314_buildGRCh37   53.983384\n",
       "..                  ...                          ...         ...\n",
       "779  chr16_79536603_T_C  All_Metal_LDSC-CORR_Neff.v2    7.348916\n",
       "780  chr2_141789968_C_T  All_Metal_LDSC-CORR_Neff.v2    7.380489\n",
       "781  chr21_37150966_C_A  All_Metal_LDSC-CORR_Neff.v2    7.389021\n",
       "782  chr1_155348517_G_C  All_Metal_LDSC-CORR_Neff.v2    7.331335\n",
       "783  chr13_77413466_A_C  All_Metal_LDSC-CORR_Neff.v2    7.375512\n",
       "\n",
       "[977 rows x 3 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "suzuki_aragam_md = pd.concat([\n",
    "    pd.read_csv(\"meta_colocid/Suzuki_n_Aragam/GCST90132314_buildGRCh37_leads.tsv\", sep=\"\\t\").assign(study=\"GCST90132314_buildGRCh37\"),\n",
    "    pd.read_csv(\"meta_colocid/Suzuki_n_Aragam/All_Metal_LDSC-CORR_Neff.v2_leads.tsv\", sep=\"\\t\").assign(study=\"All_Metal_LDSC-CORR_Neff.v2\"),\n",
    "    ])\n",
    "\n",
    "suzuki_aragam_md[\"lead1\"] = \"chr\" + suzuki_aragam_md[\"SNP\"]\n",
    "suzuki_aragam_md = suzuki_aragam_md[[\"lead1\", \"study\", \"LOG10P\"]]\n",
    "suzuki_aragam_md.rename(columns={\"LOG10P\": \"log10p1\"}, inplace=True)\n",
    "suzuki_aragam_md"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "60a6a5db",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3769\n",
      "3769\n",
      "254\n",
      "254\n",
      "2\n",
      "2\n",
      "499\n",
      "499\n",
      "159\n",
      "159\n",
      "5959\n",
      "5959\n",
      "21\n",
      "21\n",
      "6836\n",
      "6836\n"
     ]
    }
   ],
   "source": [
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "pop_dir_tag = {\n",
    "    \"EstBB\":\"Est\",\"UKBB_AFR\":\"AFR\",\"UKBB_AMR\":\"AMR\",\"UKBB_CSA\":\"CSA\",\n",
    "    \"UKBB_EAS\":\"EAS\",\"UKBB_EUR\":\"EUR\",\"UKBB_MID\":\"MID\", \"meta_EUR\":\"meta_EUR\"\n",
    "}\n",
    "\n",
    "os.makedirs(\"annotated_coloc_results/Suzuki_Aragam_coloc\", exist_ok=True)\n",
    "\n",
    "for key, _ in pop_dir_tag.items():\n",
    "    metadata = pd.read_csv(f\"annotated_meta_summaries/{key}_summary_annotated.tsv\", sep=\"\\t\")\n",
    "\n",
    "    metadata = metadata[[\"signal\", \"metabolite\", \"maf\", \"log10p\", \"nearest_genes\", \"gene_symbols\"]].drop_duplicates()\n",
    "\n",
    "    metadata = metadata.rename(columns={\"signal\":\"signal2\", \"maf\": \"maf2\", \"log10p\": \"log10p2\"})\n",
    "\n",
    "    df = pd.read_csv(f\"meta_colocid/Suzuki_n_Aragam/Suzuki_Aragam_coloc/{key}_results.tsv\", sep=\"\\t\")\n",
    "\n",
    "    df[\"study\"] = df[\"signal1\"].str.split(\"_chr\").str[0]\n",
    "\n",
    "    df = df[df[\"PP.H4\"]>=0.8]\n",
    "\n",
    "    df = df.merge(metadata, on=(\"signal2\"), how=\"left\")\n",
    "\n",
    "    print(df.shape[0])\n",
    "\n",
    "    df = df.merge(suzuki_aragam_md, on=[\"study\",\"lead1\"], how=\"left\")\n",
    "    print(df.shape[0])\n",
    "\n",
    "    df.drop_duplicates(inplace=True)\n",
    "\n",
    "    df.to_csv(f\"annotated_coloc_results/Suzuki_Aragam_coloc/{key}_Suzuki_Aragam_coloc_results.tsv\", sep=\"\\t\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.14.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
